delfin/drivers/dell_emc/vnx_block/component_handler.py

```diff
@@ -0,0 +1,490 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import six
+from oslo_log import log
+from oslo_utils import units
+
+from delfin import exception
+from delfin.common import constants
+
+LOG = log.getLogger(__name__)
+
+
+class ComponentHandler():
+
+    EMCVNX_VENDOR = 'DELL EMC'
+    RAID_GROUP_ID_PREFIX = 'raid_group_'
+
+    STATUS_MAP = {'Ready': constants.StoragePoolStatus.NORMAL,
+                  'Offline': constants.StoragePoolStatus.OFFLINE,
+                  'Valid_luns': constants.StoragePoolStatus.NORMAL,
+                  'Bound': constants.StoragePoolStatus.NORMAL}
+
+    VOL_TYPE_MAP = {'no': constants.VolumeType.THICK,
+                    'yes': constants.VolumeType.THIN}
+
+    VOL_COMPRESSED_MAP = {'no': False,
+                          'yes': True}
+
+    def __init__(self, navi_handler=None):
+        self.navi_handler = navi_handler
+
+    def get_storage(self):
+        # get storage info
+        storage = self.navi_handler.get_agent()
+        # default state is offline
+        status = constants.StorageStatus.OFFLINE
+
+        if storage:
+            status = constants.StorageStatus.NORMAL
+
+            used_cap = 0
+            free_cap = 0
+            raw_cap = 0
+            try:
+                free_map = self.handler_storage_free_capacity()
+                print('free_map=={}'.format(free_map))
+                if free_map:
+                    raw_cap = free_map.get('raw_cap')
+                    free_cap = free_map.get('free_cap')
+                    if free_cap == 0:
+                        status = constants.StorageStatus.ABNORMAL
+            except Exception:
+                LOG.error('Get storage free capacity and raw capacity failed!')
+
+            try:
+                used_map = self.handler_storage_used_capacity()
+                print('used_map=={}'.format(used_map))
+                if used_map:
+                    used_cap = used_map.get('used_cap')
+                    if used_cap == 0:
+                        status = constants.StorageStatus.ABNORMAL
+            except Exception:
+                LOG.error('Get storage used capacity failed!')
+
+            total_cap = used_cap + free_cap
+            result = {
+                'name': storage.get('node'),
+                'vendor': ComponentHandler.EMCVNX_VENDOR,
+                'model': storage.get('model'),
+                'status': status,
+                'serial_number': storage.get('serial_no'),
+                'firmware_version': storage.get('revision'),
+                'total_capacity': int(total_cap),
+                'raw_capacity': int(raw_cap),
+                'used_capacity': int(used_cap),
+                'free_capacity': int(free_cap)
+            }
+        else:
+            # If no data is returned, it indicates that there
+            # may be a problem with the network or the device.
+            # Default return OFFLINE
```

Comment:
If the data is not available, it doesn't mean storage is OFFLINE. It maybe a temporary network issue from Client. This is misleading

Reply:
Modified